python3_user_features.rst

```diff
@@ -0,0 +1,311 @@
+Useful Python 3 features for users
+==================================
+
+
+Division
+--------
+
+In Python 2, integer division is the default, so 1/2 evaluates to 0. This means
+frequently having to explicitly convert integers to floats when working with
+integer variables::
+
+    >>> int_one = 1
+    >>> int_two = 2
+    >>> int_one / int_two
+    0
+    >>> float(int_one) / int_two
+    0.5
+
+or being careful to do things like ``/ 2.`` or ``* 0.5``. In Python 3, the
+default division will yield a float, and integer division is accessed using the //
+operator::
+
+    >>> int_one / int_two
+    0.5
+    >>> int_one // int_two
+    0
+
+This makes it safer to use by default, since there there is no longer any
+implicit conversion to integers.
+
+Recursive glob
+--------------
+
+A small but very useful feature in Python 3 is the addition of a recursive
+option in the built-in :func:`~glob.glob` function. In Python 2 and 3, this
+function can be used to find all files and directories matching a certain
+pattern::
+
+    >>> import os
+    >>> import glob
+    >>> glob.glob(os.path.join('data', '*.fits'))
+    ['data/image.fits']
+
+Now let's say that the ``data`` directory now contains FITS files both
+directly in ``data`` and in sub-directories of ``data``. In Python 3, you can
+now do::
+
+    >>> import os
+    >>> import glob
+    >>> glob.glob(os.path.join('data', '**', '*.fits'), recursive=True)
+    ['data/image1.fits', 'data/subset1/a.fits', 'data/subset1/b.fits',
+     'data/subset1/c.fits', 'data/subset2/d.fits', 'data/subset2/e.fits']
+
+The ``**`` is used to indicate the point in the path at which to look for
+recursive directories, and the ``recursive=True`` option is needed to
+correctly interpret the ``**``.
+
+.. note:: We use ``os.path.join`` instead of writing out the path
+          by hand (e.g. ``data/*.fits``) to make sure that this works on
+          Windows as well as Linux and MacOS X.
+
+Matrix multiplication operator
+------------------------------
+
+Since Python 3.5, and Numpy 1.10, it is now possible to use the ``@`` operator
+to do matrix multiplication (vector product)::
+
+    >>> x = np.array([[1, 2], [3, 4]])
+    >>> y = np.array([[3, 2], [2, -1]])
+    >>> x @ y
+    array([[ 7,  0],
+           [17,  2]])
+
```

Comment:
Perhaps this needs some bigger linear algebra examples to show how much better this makes linear algebra code...